unittests/block_tests.cpp

```diff
@@ -17,7 +17,7 @@ BOOST_AUTO_TEST_CASE_TEMPLATE( block_with_invalid_tx_test, T, testers )
    auto b = main.produce_block();
 
    // Make a copy of the valid block and corrupt the transaction
-   auto copy_b = std::make_shared<signed_block>(std::move(*b));
+   auto copy_b = std::make_shared<signed_block>(b->clone());
```

Comment:
This to me is the real issue, where you get a pointer to `signed_block` and you steal the insides.

Reply:
This is the issue. Would like to make that not possible.

Reply:
I think the right solution would be to make the pointed value const: 

`using signed_block_ptr = std::shared_ptr<const signed_block>`

Reply:
We do that for `using packed_transaction_ptr = std::shared_ptr<const packed_transaction>`. Unfortunately, we currently build the `signed_block` after creation. For example look at:
https://github.com/AntelopeIO/spring/blob/78cf07f18333c2ac0853e68b4f29641cddcbdbb4/libraries/chain/block_state.cpp#L96-L96
The hardest one probably to refactor out would be:
https://github.com/AntelopeIO/spring/blob/78cf07f18333c2ac0853e68b4f29641cddcbdbb4/libraries/chain/block_state.cpp#L359-L359

Reply:
Those could `const_cast` the constness away... it would still be the safe way to do it.

Reply:
Maybe in a PR to `main`. Although, would want to avoid `const_cast` as much as possible.

Reply:
Are the only places where a block accessible via an already constructed `shared_ptr<signed_block>` is mutated in the constructor of `block_state` and `block_state_legacy`?

If so, those constructors could be slightly reworked to first construct a local temporary `std::shared_ptr<signed_block>`.

Once the mutations on the block are complete, the `block` field in the block state (which would be changed to type `std::shared_ptr<const signed_block>` can be set using the `std::shared_ptr<signed_block>`.

Reply:
I agree on avoiding the use of `const_cast`.

I'm indifferent on whether the changes to guarantee immutability of the block held in the `block_state` should only to `main`. Though maybe if I saw the extent of the changes, I would change my mind and think it is inappropriate to push to `release/1.0`

Reply:
> I agree on avoiding the use of `const_cast`.

I don't really see why? It is such a valid use that there is a `const_pointer_cast` for [shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast). 
Doesn't it make sense if you want to create an object that is immutable after creation, that you would create it directly in a `shared_ptr<const T>`, but that you'd cast away the const if there are a couple things you need to do after calling the constructor. 

The way you suggest it, the block would have to be stored somewhere else (using a `std::shared_ptr<signed_block>`, and the block member in `block_state` remaining empty until it is ready to be assigned the `std::shared_ptr<const signed_block>` using what? A cast?

Reply:
Our `FC_REFLECT` does not like the `const`. Since `signed_block_ptr` is part of `block_state` which is `FC_REFLECT`ed this doesn't work.

Reply:
I think update [unpack( Stream& s, std::shared_ptr<T>& v)](https://github.com/AntelopeIO/spring/blob/5426c8598a0ba0b9b3d6cd1e23b763aa5ba1727c/libraries/libfc/include/fc/io/raw.hpp#L201) to use `const_cast`:

```
    template<typename Stream, typename T>
    inline void unpack( Stream& s, std::shared_ptr<T>& v)
    { try {
      bool b; fc::raw::unpack( s, b );
      if( b ) {
         v = std::make_shared<T>();
         fc::raw::unpack( s, const_cast<T&>(*v) );  // const_cast as we want to be able to unpack  std::shared_ptr<const T>
      }
    } FC_RETHROW_EXCEPTIONS( warn, "std::shared_ptr<T>", ("type",fc::get_typename<T>::name()) ) }
```

Reply:
Sure we could do that and maybe even sprinkle in some more `const_cast` to make it work. But then, you have violated const in a general function. A reason I might want to use a `shared_ptr< const T>` is so that something like `unpack` does not overwrite it. If we were to tackle this, then I think the correct place would be in FC_REFLECT. There at least you know you are doing reflection serialization. 

For example, this should fail:

```cpp
bool compare(datastream<char>& ds, std::shared_ptr<const signed_block> p) {
   std::shared_ptr<signed_block> pt;
   fc::raw::unpack(ds, p); // oops wrong one
   return *pt == *p;
}
```

Reply:
If we wanted to, we could avoid the `const_cast` by unpacking into a buffer, and then creating the `shared_ptr` as in:

```
    template<typename Stream, typename T>
    inline void unpack( Stream& s, std::shared_ptr<T>& v)
    { try {
      bool b; fc::raw::unpack( s, b );
      if( b ) {
         auto dest = new std::remove_const_t<T>();
         fc::raw::unpack( s, *dest );
         v = std::shared_ptr<T>(dest);
      }
    } FC_RETHROW_EXCEPTIONS( warn, "std::shared_ptr<T>", ("type",fc::get_typename<T>::name()) ) }
```

but then you would unnecessarily lose the `std::make_shared` optimisation.

Reply:
Hmm, that is a good point about `v = std::make_shared<T>();`.

We would have to remove:
```cpp
    template<typename Stream, typename T> inline void unpack( Stream& s, const T& vi )
    {
       static_assert(not std::is_same_v<const T, const T>, "can't unpack const type");
    }

```
But seems that is not necessary. At best it seems it only provides a better error message.

Reply:
See #744